src/hotspot/share/services/threadService.hpp

```diff
@@ -108,8 +108,13 @@ class ThreadService : public AllStatic {
 
   static jlong exited_allocated_bytes()       { return _exited_allocated_bytes; }
   static void incr_exited_allocated_bytes(jlong size) {
-    // No need for atomicity, method is called under the Threads_lock
-    _exited_allocated_bytes += size;
+    // No need for an atomic add because called under the Threads_lock,
+    // but because _exited_allocated_bytes is read concurrently, need
+    // atomic store to avoid readers seeing a partial update. Also need
+    // acquire/release because the hardware might move a subsequent
+    // store to _exited_allocated_bytes above this one.
+    jlong old = Atomic::load_acquire(&_exited_allocated_bytes);
+    Atomic::release_store(&_exited_allocated_bytes, old + size);
```

Comment:
First yes Atomic::load is not needed here due to the lock, it is needed in the accessor above - sorry that wasn't clear. The Atomic store is needed of course.

But the load_acquire is definitely not needed as there is an implicit acquire when the lock is acquired, so any load of the field here must see the last value written under the lock. The release_store is also not needed because a release is for ensuring visibility of previous stores of which there are none, and has no affect on the visibility of the store done by the `release_store`.

So lets imagine a really weakly-ordered piece of hardware where all of the writing threads are completely properly synchronized by the use of the lock (else the hw is broken), but a lock-free reader can potentially see any of those writes out-of-order.  I think the only way to guarantee seeing the most recent write would be to issue a full fence() before the load. But I would be very surprised if any of our hardware actually operates in such a fashion.

Reply:
Thanks for the explanation. I reverted the acquire/release and the atomic load in incr_exited_allocated_bytes and added an atomic load in exited_allocated_bytes.